src/mlpack/methods/ann/activation_functions/hyper_sinh_function.hpp

```diff
@@ -0,0 +1,95 @@
+/**
+ * @file methods/ann/activation_functions/hyper_sinh_function.hpp
+ * @author Mayank Raj
+ *
+ * Definition and implementation of the Hyper-sinh function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_HYPER_SINH_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_ACTIVATION_FUNCTIONS_HYPER_SINH_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+
+/**
+ * The Hyper-sinh function, defined by
+ *
+ * @f{eqnarray*}{
+ * f(x) &=& \begin{cases}
+ * \sinh(x) / 3, & \text{if } x > 0 \\
+ * (x^3) / 4, & \text{if } x \leq 0
+ * \end{cases} \\
+ * f'(x) &=& \begin{cases}
+ * \cosh(x) / 3, & \text{if } x > 0 \\
+ * (3/4) x^2, & \text{if } x \leq 0
+ * \end{cases} \\
+ * @f}
+ */
+class HyperSinhFunction
+{
+ public:
+  /**
+   * Computes the Hyper-sinh function.
+   *
+   * @param x Input data.
+   * @return f(x).
+   */
+  static double Fn(const double x)
+  {
+    if (x > 0)
+      return (std::sinh(x) / 3.0);
+    else
+      return (std::pow(x, 3.0   ) / 4.0);
+  }
+
+  /**
+   * Computes the Hyper-sinh function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y)
+  {
+    y = ((x > 0) % (arma::sinh(x) / 3))
+        + ((x <= 0) % (arma::pow(x, 3) / 4));
+
+  }
+
+
+  /**
+   * Computes the first derivative of the Hyper-sinh function.
+   *
+   * @param y Input activation.
+   * @return f'(x)
+   */
+  static double Deriv(const double y)
+  {
+    if (y > 0)
+      return (std::cosh(y) / 3);
+    else
+      return ((3.0 / 4.0) * std::pow(y, 2));
+  }
+
+  /**
+   * Computes the first derivatives of the Hyper-sinh function.
+   *
+   * @param y Input activations.
+   * @param x The resulting derivatives.
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y, OutputVecType& x)
+  {
+    x = ((y > 0) % (arma::cosh(y) / 3))
```

Comment:
Did you do some benchmarks if that is more efficient, than iterating over it? In this implementation, we run both versions on each element.

Reply:
Hey initially i referred [here](https://github.com/mlpack/mlpack/blob/3b893babc1a4693c6105201d93bcf5820a6eec10/src/mlpack/methods/ann/activation_functions/elish_function.hpp#L73) as that was a similar implementation. Now when I tried to compare them the loop outperforms in terms of efficiency.

Vectored :- 0.00127024
loop :- 0.000552117

Reply:
@zoq should I keep it this way because I found few other places used similar implementation or should I use via iterating over the elements?

Reply:
If loop is faster, let's keep the loop for sure.